src/rewards/poprewards.cpp

```diff
@@ -0,0 +1,114 @@
+#include <cassert>
+#include <vector>
+#include <veriblock/entities/atv.hpp>
+#include <veriblock/rewards/poprewards.hpp>
+#include <veriblock/rewards/poprewards_calculator.hpp>
+
+namespace altintegration {
+
+static int getBestPublicationHeight(
+    const std::vector<VbkEndorsement>& endorsements,
+    const VbkBlockTree& vbk_tree) {
+  int bestPublication = -1;
+  for (const VbkEndorsement& e : endorsements) {
+    auto* block = vbk_tree.getBlockIndex(e.blockOfProof);
+    if (block == nullptr) continue;
+
+    if (block->height < bestPublication || bestPublication < 0)
+      bestPublication = block->height;
+  }
+  return bestPublication;
+}
+
+PopRewardsBigDecimal PopRewards::scoreFromEndorsements(
+    const AltBlock& endorsedBlock) {
+  PopRewardsBigDecimal totalScore = 0.0;
+
+  /// TODO: in Java we were limiting the search in containingBlocks list.
+  ///      Maybe we should do the same here as well.
+  auto endorsements = erepo_.get(endorsedBlock.getHash());
+
+  // we simply find the lowest VBK height in the endorsements
+  int bestPublication = getBestPublicationHeight(endorsements, vbk_tree_);
+  if (bestPublication < 0) return totalScore;
+
+  for (const VbkEndorsement& e : endorsements) {
+    auto* block = vbk_tree_.getBlockIndex(e.blockOfProof);
+    if (block == nullptr) continue;
+    int relativeHeight = block->height - bestPublication;
+    assert(relativeHeight >= 0);
+    totalScore += calculator_.getScoreMultiplierFromRelativeBlock(relativeHeight);
+  }
+  return totalScore;
+}
+
+PopRewardsBigDecimal PopRewards::calculateDifficulty(
+    const std::vector<AltBlock>& blocks) {
+  if (blocks.size() > rewardParams_.rewardSettlementInterval() +
+                      rewardParams_.difficultyAveragingInterval()) {
+    throw std::logic_error(
+        "amount of blocks must be less or equal than "
+        "rewardSettlementInterval + difficultyAveragingInterval");
+  }
+
+  if (blocks.size() < rewardParams_.rewardSettlementInterval()) {
+    throw std::logic_error(
+        "amount of blocks must be higher or equal than "
+        "rewardSettlementInterval");
+  }
+
+  PopRewardsBigDecimal difficulty = 0.0;
+
+  std::vector<AltBlock> sortedBlocks(blocks);
+  std::sort(
+      sortedBlocks.begin(),
+      sortedBlocks.end(),
+      [](const AltBlock& a, const AltBlock& b) { return a.height < b.height; });
```

Comment:
is it really necessary to sort?

Reply:
We have to make sure blocks are in correct order. It was made this way in Java code. We can remove sorting but enforce correct order with additional checks and sorting on the client side.

Reply:
you should use `const BlockIndex<AltBlock>& index`. 
Since there's a linked list, there's no way blocks will be in reverse order